Zephyr 3.2 Upgrade (#1499)
zmkfirmware/zmk: docs/blog/2023-03-31-zephyr-3-2.md

---
title: "Zephyr 3.2 Update"
author: Pete Johanson
author_title: Project Creator
author_url: https://gitlab.com/petejohanson
author_image_url: https://www.gravatar.com/avatar/2001ceff7e9dc753cf96fcb2e6f41110
tags: [firmware, zephyr, core]
---

I'm happy to announce that we have completed the [work](https://github.com/zmkfirmware/zmk/pull/1499) to upgrade ZMK to [Zephyr 3.2](https://docs.zephyrproject.org/3.2.0/releases/release-notes-3.2.html)!

[petejohanson] did the upgrade work to adjust ZMK for the Zephyr changes, with help from [Nicell] on the LVGL pieces.

- Upgrade to LVGL 8.x API, and move to the new Kconfig settings.
- Tons of RP2040 work.
- Zephyr core API changes, including DTS `label` use changes.
- Move to [pinctrl](https://docs.zephyrproject.org/3.2.0/hardware/pinctrl/index.html) Zephyr subsystem.

## Getting The Changes

Use the following steps to update to the latest tooling in order to properly use the new ZMK changes:

### User Config Repositories Using GitHub Actions

> Existing user config repositories using Github Actions to build will pull down Zephyr 3.2 automatically, however to build properly, if it hasn't already been changed, the repository needs to be updated to use the `stable` Docker image tag for the build:

Author: Added, if you wouldn't mind reviewing my wording. Good idea, thanks!

```
&i2c0 {
    compatible = "nordic,nrf-twi";
    sda-pin = <17>;
    scl-pin = <20>;
};
```

With the move to the `pinctrl` system, this setup now look like:

```
 &i2c0 {
    compatible = "nordic,nrf-twi";
    pinctrl-0 = <&i2c0_default>;
    pinctrl-1 = <&i2c0_sleep>;
    pinctrl-names = "default", "sleep";
 };
```

which references the `pinctrl` configuration:

```
&pinctrl {
       i2c0_default: i2c0_default {
               group1 {
                       psels = <NRF_PSEL(TWIM_SDA, 0, 17)>,
                               <NRF_PSEL(TWIM_SCL, 0, 20)>;
               };
       };

       i2c0_sleep: i2c0_sleep {
               group1 {
                       psels = <NRF_PSEL(TWIM_SDA, 0, 17)>,
                               <NRF_PSEL(TWIM_SCL, 0, 20)>;
                       low-power-enable;
               };
       };
};
```

Although slightly more _verbose_ this allows pin configuration infrastructure to be re-used, specify other modes, like sleep, etc. in a standard way across architectures.

> #### Out of Tree Boards/Shields

Reviewer: Need to mention adding `CONFIG_PINCTRL=y` to your _defconfig file.

Author: Good catch. Added.

[petejohanson] did the upgrade work to adjust ZMK for the Zephyr changes, with help from [Nicell] on the LVGL pieces.

- Upgrade to LVGL 8.x API, and move to the new Kconfig settings.
- Tons of RP2040 work.
- Zephyr core API changes, including DTS `label` use changes.
- Move to [pinctrl](https://docs.zephyrproject.org/3.2.0/hardware/pinctrl/index.html) Zephyr subsystem.

## Getting The Changes

Use the following steps to update to the latest tooling in order to properly use the new ZMK changes:

### User Config Repositories Using GitHub Actions

Existing user config repositories using Github Actions to build will pull down Zephyr 3.2 automatically, however if you created your user config a while ago, you may need to update it to reference our shared build configuration to leverage the correct Docker image.

1.  Replace the contents of your `.github/workflows/build.yaml` with:

    ```
    on: [push, pull_request, workflow_dispatch]

    jobs:
        build:
        uses: zmkfirmware/zmk/.github/workflows/build-user-config.yml@main
    ```

1.  If it doesn't exist already, add a new file to your repository named `build.yaml`:

>         ```

Reviewer: Indenting by an extra level causes this to be rendered as a code block already, so the 

    ```

at the start and end of this are appearing in the output. Either unindent this or remove the backticks.

Author: Fixed. Good catch. Thanks!

```
&i2c0 {
    compatible = "nordic,nrf-twi";
    sda-pin = <17>;
    scl-pin = <20>;
};
```

With the move to the `pinctrl` system, this setup now look like:

```
 &i2c0 {
    compatible = "nordic,nrf-twi";
    pinctrl-0 = <&i2c0_default>;
    pinctrl-1 = <&i2c0_sleep>;
    pinctrl-names = "default", "sleep";
 };
```

which references the `pinctrl` configuration:

```
&pinctrl {
       i2c0_default: i2c0_default {
               group1 {
                       psels = <NRF_PSEL(TWIM_SDA, 0, 17)>,
                               <NRF_PSEL(TWIM_SCL, 0, 20)>;
               };
       };

       i2c0_sleep: i2c0_sleep {
               group1 {
                       psels = <NRF_PSEL(TWIM_SDA, 0, 17)>,
                               <NRF_PSEL(TWIM_SCL, 0, 20)>;
                       low-power-enable;
               };
       };
};
```

Although slightly more _verbose_ this allows pin configuration infrastructure to be re-used, specify other modes, like sleep, etc. in a standard way across architectures.

#### Out of Tree Boards/Shields

> All of the in-tree boards and shields have been upgraded, but if you maintain/use and out-of-tree board or shield that uses the converted boards and overrides pins for various buses, you may need to switch to `pinctrl` to match ZMK's new approach.

Reviewer 2: ```suggestion
All of the in-tree boards and shields have been upgraded, but if you maintain/use an out-of-tree board or shield that uses the converted boards and overrides pins for various buses, you may need to switch to `pinctrl` to match ZMK's new approach.
```
Just this seems to be missing from my earlier suggestions.

The approach is the following when updating a _board_:

1. Add an entry `CONFIG_PINCTRL=y` to the `<board>_defconfig` file in the board directory.
1. Add a new file with the naming convention `<board>-pinctrl.dtsi` to your board directory.

> 1. In the new file, add your `pinctrl` entries that set up different pin control configurations for whatever peripherals/busses are needed. Here's the nice!nano file as an example:

Reviewer 2: ```suggestion
1. In the new file, add your `pinctrl` entries that set up different pin control configurations for whatever peripherals/buses are needed. Here's the nice!nano file as an example:
```
"Buses" seems to be the preferred peripheral form from some googling.

[petejohanson] did the upgrade work to adjust ZMK for the Zephyr changes, with help from [Nicell] on the LVGL pieces.

- Upgrade to LVGL 8.x API, and move to the new Kconfig settings.
- Tons of RP2040 work.
- Zephyr core API changes, including DTS `label` use changes.
- Move to [pinctrl](https://docs.zephyrproject.org/3.2.0/hardware/pinctrl/index.html) Zephyr subsystem.

## Getting The Changes

Use the following steps to update to the latest tooling in order to properly use the new ZMK changes:

### User Config Repositories Using GitHub Actions

Existing user config repositories using Github Actions to build will pull down Zephyr 3.2 automatically, however if you created your user config a while ago, you may need to update it to reference our shared build configuration to leverage the correct Docker image.

1.  Replace the contents of your `.github/workflows/build.yaml` with:

    ```
    on: [push, pull_request, workflow_dispatch]

    jobs:
        build:
        uses: zmkfirmware/zmk/.github/workflows/build-user-config.yml@main
    ```

> 1.  Add a new file to your repository named `build.yaml`:

Reviewer 2: ```suggestion
1.  If it doesn't exist already, add a new file to your repository named `build.yaml`:
```
Might be worth qualifying since there was a period of time after 3.0 upgrade that we had build.yaml but not the reusable workflow

        ```
        # This file generates the GitHub Actions matrix
        # For simple board + shield combinations, add them
        # to the top level board and shield arrays, for more
        # control, add individual board + shield combinations to
        # the `include` property, e.g:
        #
        # board: [ "nice_nano_v2" ]
        # shield: [ "corne_left", "corne_right" ]
        # include:
        #   - board: bdn9_rev2
        #   - board: nice_nano_v2
        #     shield: reviung41
        #
        ---
        ```

> and then update is as appropriate to build the right shields/boards for your configuration.

Reviewer 2: ```suggestion
and then update it as appropriate to build the right shields/boards for your configuration.
```

### Upgrade a manual script

If you have a custom GitHub Actions workflow you need to maintain for some reason, you can update the workflow to to use the `stable` Docker image tag for the build:

- Open `.github/workflows/build.yml` in your editor/IDE
- Change `zmkfirmware/zmk-build-arm:2.5` to `zmkfirmware/zmk-build-arm:stable` wherever it is found
- Locate and delete the lines for the DTS output step, which is no longer needed:

  ```
    - name: ${{ steps.variables.outputs.display-name }} DTS File
      if: ${{ always() }}
      run: |
        if [ -f "build/zephyr/${{ matrix.board }}.pre.tmp" ]; then cat -n build/zephyr/${{ matrix.board }}.pre.tmp; fi
        if [ -f "build/zephyr/zephyr.dts" ]; then cat -n build/zephyr/zephyr.dts; fi
  ```

> :::note

Reviewer 2: This note probably isn't necessary anymore given the previous section instructions, right?

Author: Agreed. Removed. I had copied from the previous Zephyr upgrade blog post.

Once the container has rebuilt, VS Code will be running the 3.2 Docker image.

### Local Host Development

The following steps will get you building ZMK locally against Zephyr 3.2:

- Run the updated [toolchain installation](/docs/development/setup#toolchain-installation) steps, and once completed, remove the previously installed SDK version (optional, existing SDK should still work)
- Install the latest version of `west` by running `pip3 install --user --update west`.
- pull the latest ZMK `main` with `git pull` for your ZMK checkout
- run `west update` to pull the updated Zephyr version and its dependencies

From there, you should be ready to build as normal!

## Windows Battery Reporting Fix

Zephyr 3.2 introduced [a new Kconfig setting](https://github.com/zephyrproject-rtos/zephyr/pull/48929) that can be used to work around a bug in Windows related to battery reporting. Check out our [bluetooth config](/docs/config/bluetooth) for the full details. The key new configuration that can be set if using Windows is:

```
CONFIG_BT_GATT_ENFORCE_SUBSCRIPTION=n
```

## Keymap Changes

> Due to conflicts with new devicetree node labels added for Zephyr's [reset system](https://example.com/TODO), the `&reset` behavior has been renamed to `&sys_reset`.

Reviewer 2: Missing link here, still

All of the in-tree keymaps have been fixed, but you may encounter build failures about duplicate names, requiring you rename the behavior reference in your keymap. Use the [Keymap Upgrader](docs/codes/keymap-upgrader) and this will get fixed for you automatically.

## Board/Shield Changes

> The following changes have [already been completed](https://github.com/zmkfirmware/zmk/pull/1143/commits) for all boards/shields in ZMK `main` branch. For existing or new PRs, or out of tree boards, the following changes are necessary to properly work with the latest changes.

Reviewer 2: ```suggestion
The following changes have [already been completed](https://github.com/zmkfirmware/zmk/pull/1499/commits) for all boards/shields in ZMK `main` branch. For existing or new PRs, or out of tree boards, the following changes are necessary to properly work with the latest changes.
```

```
&i2c0 {
    compatible = "nordic,nrf-twi";
    sda-pin = <17>;
    scl-pin = <20>;
};
```

With the move to the `pinctrl` system, this setup now look like:

```
 &i2c0 {
    compatible = "nordic,nrf-twi";
    pinctrl-0 = <&i2c0_default>;
    pinctrl-1 = <&i2c0_sleep>;
    pinctrl-names = "default", "sleep";
 };
```

which references the `pinctrl` configuration:

```
&pinctrl {
       i2c0_default: i2c0_default {
               group1 {
                       psels = <NRF_PSEL(TWIM_SDA, 0, 17)>,
                               <NRF_PSEL(TWIM_SCL, 0, 20)>;
               };
       };

       i2c0_sleep: i2c0_sleep {
               group1 {
                       psels = <NRF_PSEL(TWIM_SDA, 0, 17)>,
                               <NRF_PSEL(TWIM_SCL, 0, 20)>;
                       low-power-enable;
               };
       };
};
```

Although slightly more _verbose_ this allows pin configuration infrastructure to be re-used, specify other modes, like sleep, etc. in a standard way across architectures.

#### Out of Tree Boards/Shields

> All of the in-tree boards and shields have been upgrade, but if you maintain/use and out-of-tree shield that uses the converted boards and overrides pins for various buses, you may need to switch to `pinctrl` to match ZMK's new approach.

Reviewer 2: ```suggestion
All of the in-tree boards and shields have been upgraded, but if you maintain/use an out-of-tree board or shield that uses the converted boards and overrides pins for various buses, you may need to switch to `pinctrl` to match ZMK's new approach.
```

                           low-power-enable;
                   };
           };

           i2c0_default: i2c0_default {
                   group1 {
                           psels = <NRF_PSEL(TWIM_SDA, 0, 17)>,
                                   <NRF_PSEL(TWIM_SCL, 0, 20)>;
                   };
           };

           i2c0_sleep: i2c0_sleep {
                   group1 {
                           psels = <NRF_PSEL(TWIM_SDA, 0, 17)>,
                                   <NRF_PSEL(TWIM_SCL, 0, 20)>;
                           low-power-enable;
                   };
           };
   };
   ```

1. From the main `<board>.dts` file, add an `#include "<board>-pinctrl.dtsi"` to have the C-preprocessor combine the files.
1. Update the various peripheral nodes to use the new `pinctrl` configurations. For example, the following old configuration:

   ```
   &i2c0 {
       compatible = "nordic,nrf-twi";
       sda-pin = <15>;
       scl-pin = <17>;
   };
   ```

   would be changed to:

   ```
   &i2c0 {
       compatible = "nordic,nrf-twi";
       pinctrl-0 = <&i2c0_default>;
       pinctrl-1 = <&i2c0_sleep>;
       pinctrl-names = "default", "sleep";
   };
   ```

> Because `pinctrl` configuration is very dependant on the specific target SoC, you will rarely need to consider it for a shield overlay that leverages a pro micro or XIAO abstraction. As noted, you're more likely to need to fix up pinctrl settings is using a board specific shield overlay, e.g. `<shield>/boards/<board>.overlay` to set things up.

Reviewer 2: ```suggestion
Because `pinctrl` configuration is very dependent on the specific target SoC, you will rarely need to consider it for a shield overlay that leverages a pro micro or XIAO abstraction. As noted, you're more likely to need to fix up pinctrl settings is using a board specific shield overlay, e.g. `<shield>/boards/<board>.overlay` to set things up.
```
